### PR TITLE
include model and dataset hash update and upload in retrain pipeline

### PR DESCRIPTION
**Pull request by ruinianxu**
# Description

This PR is for further automating hitl retraining pipeline with including parameter sweep, sending training jobs to cluster, picking the best model, computing the hash for model and dataset, and lastly compressing and uploading to AWS.

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Before: nsp_retrain_infra.py only support sending multiple jobs to cluster and choose the best model after training is done.
After: after picking up the best model, the script will move the best model to the corresponding folder of artifacts. Besides, log file for the best model will be also placed under the folder. Then it computes hash for models and datasets and compress and upload to AWS.

# Testing

Please describe the tests that you ran to verify your changes and how we can reproduce:

For the purpose of testing, please use batch id of 20220524_1653470045. Before running retraining script, you will need to run the following script to generate split mask of the old data. We consider all data to be new here. Move the generated split_masks.pth under /fairo/droidlet/tools/hitl/nsp_retrain/configs.

```
import torch
# start and end index for train, val and test split
train_s_idx = 0
train_e_idx = 0
val_s_idx = 0
val_e_idx = 0
test_s_idx = 0
test_e_idx = 0
# generate list
train_list = [i for i in range(train_s_idx, train_e_idx)]
val_list = [i for i in range(val_s_idx, val_e_idx)]
test_list = [i for i in range(test_s_idx, test_e_idx)]
# turn into tensor
train_list = torch.tensor(train_list)
val_list = torch.tensor(val_list)
test_list = torch.tensor(test_list)
save_path = "./split_masks.pth"
data = {"annotated": {
"train": train_list,
"valid": val_list,
"test": test_list,
}}
torch.save(data, save_path)
```

For quick test which only train for 2 epochs, run
```
python droidlet/tools/hitl/nsp_retrain_infra.py --batch_id 20220524_1653470045 --retrain_data_splits 2 2 2 --num_epochs 2 -decoder_learning_rate .0000005 .000001 .000005 -encoder_learning_rate .0 .000001
```

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [x] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

**Review by kavyasrinet — COMMENTED** [review]
Looks good to me but I'll default to @snyxan 's approval here.

File: droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py [hunk 1]
```diff
@@ -14,6 +14,82 @@
 print("Rootdir : %r" % ROOTDIR)
 
 
+def tar_and_upload(agent, artifact_name, model_name=None):
```
Comment (kavyasrinet) [inline droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:17]: I don't understand - why do we still have another function to do this ?

Comment (kavyasrinet) [reply droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:17]: Instead of adding a new function that redundantly does a part of `compute_checksum_tar_and_upload`, can we extract out the shared functionality of the two functions (actually tar'ing and uploading) and create a new method that both these methods call ?

File: droidlet/tools/hitl/nsp_retrain/README.md [hunk 2]
```diff
@@ -54,4 +54,10 @@ Note that we are providing the following parameters:
 you can also provide environment variables mentioned in step#1 here by running command like:
 ```
 MEPHISTO_AWS_ACCESS_KEY_ID="xxxxx" MEPHISTO_AWS_SECRET_ACCESS_KEY="xxxx" .... python main.py --interaction_job_num 2 --image_tag <TAG_NAME> --task_name <TASK_NAME> --resample
-```
\ No newline at end of file
+```
+
```
Comment (kavyasrinet) [inline droidlet/tools/hitl/nsp_retrain/README.md:58]: Thanks for adding this, this is super useful

**Review by kavyasrinet — COMMENTED** [review]
I requested a refactor like change in the artifact script, but other than that , looks good.

File: droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py (hunk 1 above)
Comment (ruinianxu) [reply droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:17]: It is a good idea. I will refactor ```compute_checksum_tar_and_upload```.

Comment (ruinianxu) [reply droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py]: @kavyasrinet 
I've refactored the previous method ```compute_checksum_tar_and_upload```, which consists of mainly two parts (1.compute checksum, 2.tar and upload). Since computing checksum has been implemented in ```compute_checksum.py```, I directly import from it. Then I move tar and upload out as an independent method. Lastly, I rename ```compute_checksum_tar_and_upload``` to ```upload_artifacts_to_aws```. Please let me know if it is reasonable.

File: droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py [hunk 3]
```diff
@@ -14,45 +14,32 @@
 print("Rootdir : %r" % ROOTDIR)
 
 
-def compute_checksum_tar_and_upload(agent, artifact_name, model_name=None):
+def tar_and_upload(agent, artifact_name, model_name=None, checksum=None):
+    """
+    Tar all files for the artifact and upload it to AWS.
+    """
     if not agent:
```
Comment (kavyasrinet) [inline droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:21]: I'm not sure why we need to do this step all over again ? `compute_checksum` already does this for us. We can just start from :
```
print("Now making the tar file...")
process = Popen(...
```
right ?

File: droidlet/tools/artifact_scripts/compute_checksum.py [hunk 4]
```diff
@@ -56,9 +58,12 @@ def compute_checksum_for_directory(agent=None, artifact_type=None, model_name=No
         checksum_name = "datasets.txt"
 
     artifact_path = os.path.join(ROOTDIR, "droidlet/artifacts", artifact_folder_name)
-    checksum_write_path = os.path.join(
-        ROOTDIR, "droidlet/tools/artifact_scripts/tracked_checksums/" + checksum_name
-    )
+    if checksum_write_path is None:
```
[the remark this answers is not in the document]
Comment (ruinianxu) [reply droidlet/tools/artifact_scripts/compute_checksum.py]: I think this commit is out-of-date right now. After yesterday's meeting, I converted this function back. Right now there is no changes. 

Previously, I added ```checksum_write_path``` since I imported it for ```compute_checksum_tar_upload``` which stores hash in checksum.txt. The default ```compute_checksum``` method only saves hash to files under ```tracked_checksums```. ```checksum_write_path``` specifies where to the computed hash.

File: droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py (hunk 3 above)
Comment (kavyasrinet) [reply droidlet/tools/artifact_scripts/upload_artifacts_to_aws.py:21]: My point is just give these as input to the function rather than recomputing artifact_path_name

**Review by kavyasrinet — APPROVED** [review]
Thanks for refactoring in the comments, LGTM
